### PR TITLE
HasPlatformType rule

### LINKED ISSUE #556: Rule: Require explicit types when using platform types
I watched "[Advanced Kotlin](http://www.oredev.org/2017/sessions/retrofitting-with-retrofit-tips-on-updating-an-old-networking-layer)" talk (relevant discussion starts around 17:00 in the linked video), and have an idea for a new check for when platform types are used. Example taken from [here](https://github.com/christinalee/oredev17/blob/d4349605518f858cc65f5445e09edba10fae7ead/src/1_Types.kt#L719-L776):

If you have this in Java (with no nullability annotations):
```
import java.util.ArrayList;
import java.util.List;

public class PlatformTypes {
    public static List<String> returnsNullList() {
        return null;
    }
}
```
and you call it from Kotlin:
```
fun errorWithoutExplicitType() {
    val list = returnsNullList()
    list.size
}
```
Then `list.size` throws an NPE because `list` is `null`, but this was unknown at compile time. The recommendation is to explicitly set the type on the declaration:
```
fun errorWithExplicitType() {
    val list: List<String> = returnsNullList()
    list.size
}
```
This way the error changes from NPE on `list.size` to an IllegalStateException at `val list: List<String> = returnsNullList()` with description like "list must not be null". This helps with debugging because the exception occurs when the value is assigned to the `list` property, not when `list.size` is called, which could happen several lines down. This wastes a developer's time when debugging because they have to trace back to the location `list` was declared.

detekt should check that when a property is assigned a value, and the expression resolves to a platform type, that the property has an explicit type, to improve the experience when working with platform types.

This also requires type and symbol resolving :)

### PR DESCRIPTION
**Pull request by 3flex**
Closes #556 

Rule is quite conservative and picks up usage in public APIs only. Future enhancement could be to add a config option to detect implicit usage of platform types anywhere they're found.

### CODE REVIEW

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/HasPlatformType.kt [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+import org.jetbrains.kotlin.psi.KtProperty
+import org.jetbrains.kotlin.psi.psiUtil.containingClassOrObject
+import org.jetbrains.kotlin.resolve.BindingContext
+import org.jetbrains.kotlin.types.isFlexible
+
+/**
+ * Platform types must be declared explicitly in public APIs to prevent unexpected errors.
+ *
+ * Based on code from Kotlin project:
+ * https://github.com/JetBrains/kotlin/blob/1.3.50/idea/src/org/jetbrains/kotlin/idea/intentions/SpecifyTypeExplicitlyIntention.kt#L86-L107
```
Comment (schalkms) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/HasPlatformType.kt]: I'm not sure if we should really reference the source location.
Source code can be changed easily. Thus, this reference could become out of date.
It's better IMHO to include the license in our project.

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/HasPlatformType.kt [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        Severity.Maintainability,
+        "Platform types must be declared explicitly in public APIs.",
+        Debt.FIVE_MINS
+    )
+
+    override fun visitKtElement(element: KtElement) {
+        super.visitKtElement(element)
+
+        if (bindingContext == BindingContext.EMPTY) return
+        if (element is KtCallableDeclaration && element.hasImplicitPlatformType()) {
```
Comment (schalkms) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/HasPlatformType.kt]: The ifs can be merged easily.
```suggestion
        if (bindingContext != BindingContext.EMPTY && element is KtCallableDeclaration && element.hasImplicitPlatformType()) {
```

File: detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/HasPlatformTypeSpec.kt [hunk 3]
```diff
@@ -0,0 +1,74 @@
+package io.gitlab.arturbosch.detekt.rules.bugs
+
+import io.gitlab.arturbosch.detekt.api.Config
+import io.gitlab.arturbosch.detekt.test.KtTestCompiler
+import io.gitlab.arturbosch.detekt.test.compileAndLintWithContext
+import org.assertj.core.api.Assertions.assertThat
+import org.spekframework.spek2.Spek
+import org.spekframework.spek2.style.specification.describe
+
+object HasPlatformTypeSpec : Spek({
```
Comment (schalkms) [inline detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/HasPlatformTypeSpec.kt]: Could we please make sure that the tests cover all branches (if possible)?

```Kotlin
if (containingClassOrObject?.isLocal == true) return false

val callable = bindingContext[BindingContext.DECLARATION_TO_DESCRIPTOR, this] as? CallableDescriptor ?: return false
if (!callable.visibility.isPublicAPI) return false
return callable.returnType?.isFlexible() ?: return false
```

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/HasPlatformType.kt (hunk 1 above)
Comment (schalkms) [reply detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/HasPlatformType.kt]: relates to #1917

Comment (3flex) [reply detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/HasPlatformType.kt]: The url is fixed to v1.3.50 of Kotlin, so the link won't break.

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/HasPlatformType.kt [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+            is KtProperty -> if (isLocal || typeReference != null) return false
+            else -> return false
+        }
+
+        if (containingClassOrObject?.isLocal == true) return false
+
+        val callable =
+            bindingContext[BindingContext.DECLARATION_TO_DESCRIPTOR, this] as? CallableDescriptor ?: return false
+        if (!callable.visibility.isPublicAPI) return false
+        return callable.returnType?.isFlexible() ?: return false
```
[the remark this answers is not in the document]
Comment (3flex) [reply detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/HasPlatformType.kt]: https://kotlin.github.io/kotlin-spec/#flexible-types

**Pull request merged**